File: cpp/LeetCode/No279.cpp
/*
Given an integer n, return the least number of perfect square numbers that sum to n.

A perfect square is an integer that is the square of an integer; in other words, it is the product of some integer with itself. 
For example, 1, 4, 9, and 16 are perfect squares while 3 and 11 are not.
*/
#include <iostream>
#include <vector>
#include <climits>
#include <cmath>
using namespace std;

class Solution {
public:
    int numSquares(int n) {
        vector<int> mem(n+1, -1);
        return dp(n, mem);
    }
private:
    int dp(int n, vector<int> &mem){
        if(n==0) return 0;
        if(mem[n]!=-1) return mem[n];

        int res = INT_MAX;
        for(int i=1;i<=n;i++){
            int j = i*i;
            if(j<=n){
                res = min(res, dp(n-j, mem)+1);
            }
            else{
                break;
            }
        }
        return mem[n] = res;
    }
};

int main(void){
    Solution solution;
    int n;

    n = 12;
    cout << solution.numSquares(n) << endl;

    n = 13;
    cout << solution.numSquares(n) << endl;
    return 0;
}

File: python/LeetCode Quiz/No279.py
'''
Given an integer n, return the least number of perfect square numbers that sum to n.

A perfect square is an integer that is the square of an integer; 
in other words, it is the product of some integer with itself. 
For example, 1, 4, 9, and 16 are perfect squares while 3 and 11 are not.
'''

class Solution:
    def numSquares(self, n: int) -> int:
        nodes = []
        for i in range(1, n+1):
            if i**2<=n:
                nodes.append(i**2)
            else:
                break

        q = [0]
        mem = set()
        count = 0
        while q:
            size = len(q)
            count+=1
            for _ in range(size):
                num = q.pop(0)
                for val in nodes:
                    temp_sum = num+val
                    if temp_sum == n:
                        return count
                    if temp_sum not in mem and temp_sum<n:
                        q.append(temp_sum)
                        mem.add(temp_sum)
            
        return -1

def main():
    solution = Solution()
    
    n = 12
    print(solution.numSquares(n))

if __name__=='__main__':
    main()
